src/basic/env-util.c

```diff
@@ -16,22 +16,26 @@
 #include "strv.h"
 #include "utf8.h"
 
-#define VALID_CHARS_ENV_NAME                    \
+/* We follow bash for the character set. Different shells have different rules. */
+#define VALID_BASH_ENV_NAME_CHARS               \
         DIGITS LETTERS                          \
         "_"
 
-static bool env_name_is_valid_n(const char *e, size_t n) {
-        const char *p;
+static bool printable_portable_character(char c) {
+        /* POSIX.1-2008 specifies almost all ASCII characters as "portable". (Only DEL is excluded, and
+         * additionally NUL and = are not allowed in variable names). We are stricter, and additionally
+         * reject BEL, BS, HT, CR, LF, VT, FF and SPACE, i.e. all whitespace. */
+
+        return c >= '!' && c <= '~';
+}
 
+static bool env_name_is_valid_n(const char *e, size_t n) {
         if (!e)
                 return false;
 
         if (n <= 0)
                 return false;
 
-        if (e[0] >= '0' && e[0] <= '9')
-                return false;
```

Comment:
urks. this is a security problem... Why this? did anyone ask for this?

Please, let's sanitize data. This is expressly mentioned in POSIX to be problematic and noone asked so far to allow this, so why change this?